revise internal use and format of device names. Fixes #1320

phillxnet opened the pull request:
Changes the format of disk names saved in db Disk.name field from sda type to by-id type.
Builds on existing conversion methods established in pr #1292 to translate transient sda type names to by-id type names.

It was found that in some corner cases such as virtio drives with no serial that by-id type names were not available. In this case the vda type name is stored instead. However this creates an issue with assumptions on drive path now being /dev/disk/by-id from the previous /dev type path. As this affected smart calls via the wrapper of all smart calls introduced in #1206 to find the base device and the enhancement of this introduced in #1221 we simply disable smart calls on devices that have been found to have no serial ie ascribed with fake-serial-. Also to account for this possible existence in the Disk.name field of devices that don't have by-id type names these devices have also been excluded in all future pool creation by adding a check for fake-serial to disk availability for pool selection. This avoids future pools being created with disks that we can't track due to no serial number and hence no by-id type name. The existing user facing serial warning has been updated to indicate this.

Note that as the structure of by-id type names affords a syntactic method for acquiring the base device it was also possible to remove system calls to lsblk to establish these base devices for use in all SMART system calls. As only transient use is made of sda type names, mainly when reading directly from lsblk in scan_disks() and from the immediate output of some btrfs commands and given all internal uses are now transitioned to by-id type we have an enhancement in log diagnostic capability as all calls involving drive names are now directly traceable to the device given by-id type names contain model and serial number and are reboot safe. On this point many more modern drives exhibit multiple by-id type names and to stabilise this I have chosen to use the longer of however many are found. This has so far proven to be the one containing model and serial rather than the shorter wwn-. This also addresses an observed and rare instability in by-id name presentation order as an earlier implementation of by-id name retrieval such as that in pr #1292 simply used the first instance listed.

An important element assumed in this pr is that no device without a reboot safe by-id type name can from here on constitute a component of new pools. This move improves code clarity and give us a sound datum on which to base our device naming.

All current disk operations have been tested as working with this pr, just prior to re-basing on 3.8-14.

Also note that a significant number of methods / functions have been moved from fs/btrfs to system/osi. This was by way of existing TODO:'s (added in #1292) and significantly reduces the possibility of circular dependencies by reducing the number of import statements across these two files in particular. The main advantage currently being the segregation of all directly btrfs related functions. Many fresh builds have proven the imports as presented in this pr to be sound (prior to re-basing). There is further improvement to be had separation wise such as setting up an hdparm specific functions file in for example system/hdparm.py and moving appropriate methods from system/osi to there, but this requires additional TODO:'s for review purposes and are included later-on in this pr for implementation in a future pr. This is in response the the same suggestion from @schakrava in pr #1292 

Known issues:
A cosmetic side effect given the by-id names are much longer is that the Disk Widget gets them all on top of one another when more than 2 disks are active. Considering this to be worthy of another issue / pr as essential function is there still and a proper fix may be to establish an index type arrangement in the widget or redo the top part of the widget some how.

Fixes #1320 and by a side effect also:
Fixes #1051 as the preferred (longest) by-id device names contain the device serial number.
It is also assumed that this pr will sort out the root cause of #1043 "dev name change breaking mounts" but this will require additional testing before confirmation.

Please see the previous version of this pr (#1338) opened 20 days before this one for developmental history to date. All currently relevant info is replicated in this pr.

phillxnet commented:
@schakrava Rebased on latest master (3.8-14 git tag) and re-tested as requested.
On regular, bios raid, and mdraid system disk installs all disk functions were as expected.
Submitting as ready for review.

MFlyer commented:
To @schakrava - Cross reference to #1354 - Both PR have mods over data_collector

schakrava commented on src/rockstor/fs/btrfs.py:
```diff
@@ -782,346 +809,25 @@ def device_scan():
     return o[0]
 
 
 def btrfs_importable(disk):
```
even though this is not a function you introduced, it's misleading and not used. Remove it if you can please.

schakrava commented on src/rockstor/storageadmin/models/disk.py:
```diff
@@ -18,14 +18,18 @@
+    Revised to contain device names for use with the udev created links at
+    /dev/disk/by-id/ which in turn are symlinks to sda, sdb etc.
+    eg ata-QEMU_HARDDISK_QM00005 ie mostly derived from model and serial number.
+    """
```
I support this change, but strongly prefer that primary user facing name of a Disk should be the more readable/humanized one such as sda, sdb etc.. Since they are not persistent, it makes sense to store the persistent by-id name in the db instead. However, I'd like to suggest a property that retrieves the readable name during runtime as the model is feteched.

This also requires some UI changes, wherein the readable name is presented to the admin for various management matters like creating a Pool, adding disks etc.., with the Disk.name in an additional column or in a tooltip. This way, we achieve a pleasant user experience while not compromising the true identity of the Disk.

schakrava commented on src/rockstor/system/osi.py:
```diff
@@ -786,48 +1234,136 @@ def set_disk_spindown(device, spindown_time, apm_value,
+    was encountered by run_command or no by-id type names were encountered.
+    """
+    byid_name_map = {}
+    out, err, rc = run_command([LS, '-l', '/dev/disk/by-id'],
```
just a recommendation -- instead of using run_command here, we can simply use python to generate this map. This entire function reduces to

import os, os.path
dirname = '/dev/disk/by-id'
m = {}
for os.listdir(dirname):
    m[os.path.split(os.path.realpath(os.path.join(dirname, d)))[-1]] = d

phillxnet commented:
@schakrava Thanks for the review.

Since you review I have removed the unused function in fs/btrfs that you indicated along with another I found in the same file. I also removed the comment inconsistency in the other function you commented on. Alas I am unable to translate it to the neater os.listdir as there was a requirement to have a 'long' list via 'ls -l' command. But this did help with my development of the suggested run time property extension of the Disk model to provide short name (ie sda type) retrieval via this extension. I have called this property temp_name to emphasise their temporary nature. The function of this property has been tested via logging prior to commit and works as expected.

As discussed outside this thread I have not enabled / re-surfaced the short names to the UI as that is best done by others more able (and quicker) in that area and is arguably better done is another pr given the vintage and size of this pr. So in the interests of getting this one out I am leaving the long and arguably cumbersome by-id type names in the UI. This mostly affects the readability of the disk names as given in the Dashboard Disk Activity widget as indicated earlier in this pr.

Given the above caveats this pr is again ready for review / merge.

@MFlyer the only additional files I have altered that impact on your large pending pr are serializers.py but I don't think this will cause any merge issues as our changes were in different areas and relatively small.

Sorry I dragged this one out folks and hope it helps.

MFlyer commented:
hi @phillxnet concurrent mods actually are on serializers and data_collect, but I totally agree with you about differents areas not making merge issues :+1: 

I'm really curious to see latest changes by you PR :)

schakrava commented:
Thank you @phillxnet !

phillxnet commented:
@schakrava Cheers.
